File: Ch06/BookStore/BookStore/ViewController.m
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

#import "ViewController.h"
#import "AppDelegate.h"
#import "BookCategory.h"
#import "Book.h"
#import "Page.h"

@interface ViewController ()

@end

@implementation ViewController

- (NSURL *)applicationDocumentsDirectory {
  return [[[NSFileManager defaultManager] URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] lastObject];
}

- (NSManagedObjectContext*)managedObjectContext {
  AppDelegate *ad = [UIApplication sharedApplication].delegate;
  return ad.managedObjectContext;
}

- (void)saveContext {
  NSManagedObjectContext *managedObjectContext = self.managedObjectContext;
  if (managedObjectContext != nil) {
    NSError *error = nil;
    if ([managedObjectContext hasChanges] && ![managedObjectContext save:&error]) {
      NSString *message = [self validationErrorText:error];
      NSLog(@"Error: %@", message);
    }
  }
}

- (void)viewDidAppear:(BOOL)animated {
  [super viewDidAppear:animated];
//  [self populateAuthors];
  [self showExampleData];
}

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view, typically from a nib.
}

- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
  // Dispose of any resources that can be recreated.
}

- (void)initStore {
  NSFileManager *fm = [NSFileManager defaultManager];
  
  NSString *seed = [[NSBundle mainBundle] pathForResource: @"seed" ofType: @"sqlite"];
  NSURL *storeURL = [[self applicationDocumentsDirectory] URLByAppendingPathComponent:@"BookStoreEnhanced.sqlite"];
  
  if([fm fileExistsAtPath:[storeURL path]]) {
    [fm removeItemAtPath:[storeURL path] error:nil];
  }
  
  if (![fm fileExistsAtPath:[storeURL path]]) {
    NSLog(@"Using the original seed");
    NSError *error = nil;
    if (![fm copyItemAtPath:seed toPath:[storeURL path] error:&error]) {
      NSLog(@"Error seeding: %@", error);
      return;
    }
    
    NSLog(@"Store successfully initialized using the original seed");
  }
  else {
    NSLog(@"The original seed isn't needed. There is already a backing store.");
    
    // Update 1
    {
      NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"Publication"];
      request.predicate = [NSPredicate predicateWithFormat:@"title=%@", @"The fourth book"];
      NSUInteger count = [self.managedObjectContext countForFetchRequest:request error:nil];
      if(count == 0) {
        NSLog(@"Applying batch update 1");
        
        NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Category"];
        NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
        
        BookCategory *category = [categories lastObject];
        Book *book4 = [NSEntityDescription insertNewObjectForEntityForName:@"Publication" inManagedObjectContext:self.managedObjectContext];
        book4.title = @"The fourth book";
        book4.price = 12;
        
        [category addBooks:[NSSet setWithObjects:book4, nil]];
        
        [self.managedObjectContext save:nil];
        
        NSLog(@"Update 1 successfully applied");
      }
    }
  }
  /* Code to add the fourth book */
  /*
   NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"BookCategory"];
   NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
   BookCategory *category = [categories lastObject];
   
   Book *book4 = [NSEntityDescription insertNewObjectForEntityForName:@"Book" inManagedObjectContext:self.managedObjectContext];
   book4.title = @"The fourth book";
   book4.price = 12;
   
   [category addBooks:[NSSet setWithObjects:book4, nil]];
   
   [self saveContext];
   */
}

- (void)insertSomeData {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"BookCategory"];
  NSArray *categories = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];

  AppDelegate *ad = [UIApplication sharedApplication].delegate;
  if(ad.unstable) {
    NSLog(@"The app is unstable. Preventing updates.");
    return;
  }
  
  BookCategory *category = [categories lastObject];
  
  // Tell the undo manager that from now on, we manage grouping
  [self.managedObjectContext.undoManager setGroupsByEvent:NO];
  
  for(int i=5; i<10; i++) {
    // Start a new group
    [self.managedObjectContext.undoManager beginUndoGrouping];
    
    Book *book = [NSEntityDescription insertNewObjectForEntityForName:@"Publication" inManagedObjectContext:self.managedObjectContext];
    book.title = [NSString stringWithFormat:@"The %dth book", i];
    book.price = i;
    [category addBooksObject:book];
    
    // End the current group
    [self.managedObjectContext.undoManager endUndoGrouping];
  }
  
  [self.managedObjectContext.undoManager undo];
  [self saveContext];
}

- (void)showExampleData {
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Publication"];
  NSArray *books = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  NSLog(@"Size: %lu", [books count]);
  for (Book *book in books) {
    NSLog(@"Title: %@, price: %.2f", book.title, book.price);
  }
}

- (NSString *)validationErrorText:(NSError *)error {
  // Create a string to hold all the error messages
  NSMutableString *errorText = [NSMutableString stringWithCapacity:100];
  // Determine whether we're dealing with a single error or multiples, and put them all
  // in an array
  NSArray *errors = [error code] == NSValidationMultipleErrorsError ? [[error userInfo] objectForKey:NSDetailedErrorsKey] : [NSArray arrayWithObject:error];
  
  // Iterate through the errors
  for (NSError *err in errors) {
    // Get the property that had a validation error
    NSString *propName = [[err userInfo] objectForKey:@"NSValidationErrorKey"];
    NSString *message;
    
    // Form an appropriate error message
    switch ([err code]) {
      case NSValidationNumberTooSmallError:
        message = [NSString stringWithFormat:@"%@ must be at least $15", propName];
        break;
      default:
        message = @"Unknown error. Press Home button to halt.";
        break;
    }
    
    // Separate the error messages with line feeds
    if ([errorText length] > 0) {
      [errorText appendString:@"\n"];
    }
    [errorText appendString:message];
  }
  return errorText;
}

- (void)populateAuthors {
  // 1. Create a list of author names to assign
  NSArray *authors = @[@"John Doe", @"Jane Doe", @"Bill Smith", @"Jack Brown"];
  
  // 2. Get all the publications from the data store
  NSFetchRequest *fetchRequest = [NSFetchRequest fetchRequestWithEntityName:@"Publication"];
  NSArray *books = [self.managedObjectContext executeFetchRequest:fetchRequest error:nil];
  for(int i=0; i<books.count; i++) {
    Book *book = books[i];
    book.price = 20+i;
    
    // 3. Set the author using one of the names in the array we created
    [book setValue:authors[i % authors.count] forKey:@"author"];
  }

  // 4. Commit everything to the store
  [self saveContext];
}

@end

File: Ch06/BookStoreSwift/BookStoreSwift/ViewController.swift
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

import UIKit
import CoreData

class ViewController: UIViewController {
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        if let managedObjectContext = self.managedObjectContext {
//            self.populateAuthors()
            self.showExampleData()
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    lazy var managedObjectContext: NSManagedObjectContext? = {
        let appDelegate = UIApplication.sharedApplication().delegate as AppDelegate
        return appDelegate.managedObjectContext
    }()
    
    func saveContext() {
        var error: NSError? = nil
        if let managedObjectContext = self.managedObjectContext {
            if managedObjectContext.hasChanges && !managedObjectContext.save(&error) {
                let message = validationErrorText(error!)
                println("Error: \(message)")
            }
        }
    }
    
    func initStore() {
        let fm = NSFileManager.defaultManager()
        
        let seed = NSBundle.mainBundle().pathForResource("seed", ofType: "sqlite")
        if let seed = seed {
            let storeURL = AppDelegate.applicationDocumentsDirectory.URLByAppendingPathComponent("BookStoreEnhancedSwift.sqlite")
            
            // Remove the old store to force re-seeding
            if fm.fileExistsAtPath(storeURL.path!) {
                fm.removeItemAtPath(storeURL.path!, error: nil)
            }
            
            if !fm.fileExistsAtPath(storeURL.path!) {
                println("Using the original seed")
                var error: NSError? = nil
                if !fm.copyItemAtPath(seed, toPath: storeURL.path!, error: &error) {
                    println("Seeding error: \(error?.localizedDescription)")
                    return
                }
                println("Store successfully initialized using the original seed.")
            }
            else {
                println("The seed isn't needed. There is already a backing store.")
            }
        }
        else {
            println("Could not find the seed.")
        }
        
        // Update 1
        let fetchRequest1 = NSFetchRequest(entityName: "Publication")
        fetchRequest1.predicate = NSPredicate(format: "title=%@", argumentArray: ["The fourth book"])
        if self.managedObjectContext?.countForFetchRequest(fetchRequest1, error: nil) == 0 {
            println("Applying batch update 1")
            let fetchRequest = NSFetchRequest(entityName: "BookCategory")
            let categories = self.managedObjectContext?.executeFetchRequest(fetchRequest, error: nil)
            
            let category = categories?.last as BookCategory
            var book4 = NSEntityDescription.insertNewObjectForEntityForName("Publication", inManagedObjectContext: self.managedObjectContext!) as Book
            book4.title = "The fourth book"
            book4.price = 12
            
            var booksRelation = category.mutableSetValueForKeyPath("books")
            booksRelation.addObject(book4)
            
            saveContext()
            println("Update 1 successfully applied")
        }
    }
    
    func insertSomeData() {
        let ad = UIApplication.sharedApplication().delegate as? AppDelegate
        
        if let unstable = ad?.unstable {
            if unstable {
                println("The app is unstable. Preventing updates.")
                return
            }
        }
  
        let category = self.managedObjectContext?.executeFetchRequest(NSFetchRequest(entityName: "BookCategory"), error: nil)?.last as BookCategory?
        
        if let managedObjectContext = self.managedObjectContext {
            managedObjectContext.undoManager?.groupsByEvent = false
            
            if let category = category {
                for i in 5..<10 {
                    managedObjectContext.undoManager?.beginUndoGrouping()
                    
                    var book = NSEntityDescription.insertNewObjectForEntityForName("Publication", inManagedObjectContext: managedObjectContext) as Book
                    book.title = "The \(i)th book"
                    book.price = Float(i)
                    
                    var booksRelation = category.mutableSetValueForKeyPath("books")
                    booksRelation.addObject(book)
                    
                    managedObjectContext.undoManager?.endUndoGrouping()
                }
                
                managedObjectContext.undoManager?.undo()

                saveContext()
            }
        }
    }
    
    func showExampleData() {
        let fetchRequest = NSFetchRequest(entityName: "Publication")
        let books = self.managedObjectContext?.executeFetchRequest(fetchRequest, error: nil)
        for book in books as [Book] {
            println(String(format: "Title: \(book.title), price: %.2f",  book.price))
        }
    }

    func populateAuthors() {
        // 1. Create a list of author names to assign
        let authors = ["John Doe", "Jane Doe", "Bill Smith", "Jack Brown"]
        
        // 2. Get all the publications from the data store
        let fetchRequest = NSFetchRequest(entityName: "Publication")
        let books = self.managedObjectContext?.executeFetchRequest(fetchRequest, error: nil)
        if let books = books {
            for var i = 0; i<books.count; i++ {
                var book = books[i] as Book
                book.price = 20 + Float(i)
                // 3. Set the author using one of the names in the array we created
                book.setValue(authors[i % authors.count], forKeyPath: "author")
            }
        }
        
        // 4. Commit everything to the store
        saveContext()
    }
    
    func validationErrorText(error : NSError) -> String {
        // Create a string to hold all the error messages
        let errorText = NSMutableString(capacity: 100)
        // Determine whether we're dealing with a single error or multiples, and put them all
        // in an array
        let errors : NSArray = error.code == NSValidationMultipleErrorsError ? error.userInfo?[NSDetailedErrorsKey] as NSArray : NSArray(object: error)
        
        // Iterate through the errors
        for err in errors {
            // Get the property that had a validation error
            let e = err as NSError
            let info = e.userInfo
            let propName : AnyObject? = info!["NSValidationErrorKey"]
            var message : String?
            
            // Form an appropriate error message
            switch err.code {
            case NSValidationNumberTooSmallError:
                message = "\(propName!) must be at least $15"
            default:
                message = "Unknown error. Press Home button to halt."
            }
            
            // Separate the error messages with line feeds
            if errorText.length > 0 {
                errorText.appendString("\n")
            }
            
            errorText.appendString(message!)
        }
        
        return errorText
    }
}

